finality-aleph/src/data_io.rs

```diff
@@ -13,73 +13,256 @@ use futures::{
+pub enum AlephData<B: BlockT> {
+    Empty,
+    HeadProposal(AlephProposal<B>),
 }
 
-impl<H, N> AlephData<H, N> {
-    pub(crate) fn new(block_hash: H, block_number: N) -> Self {
-        AlephData {
-            hash: block_hash,
-            number: block_number,
+// Need to be implemented manually, as deriving does not work (`BlockT` is not `Hash`).
+impl<B: BlockT> Hash for AlephData<B> {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        match self {
+            AlephData::Empty => {
+                (0u8).hash(state);
+            }
+            AlephData::HeadProposal(proposal) => {
+                (1u8).hash(state);
+                proposal.hash(state);
+            }
+        }
+    }
+}
+
+// Clippy does not allow deriving PartialEq when implementing Hash manually
+impl<B: BlockT> PartialEq for AlephData<B> {
+    fn eq(&self, other: &Self) -> bool {
+        match (self, other) {
+            (AlephData::Empty, AlephData::Empty) => true,
+            (AlephData::HeadProposal(p1), AlephData::HeadProposal(p2)) => p1.eq(p2),
+            _ => false,
         }
     }
 }
 
-pub(crate) type AlephDataFor<B> = AlephData<<B as BlockT>::Hash, NumberFor<B>>;
+impl<B: BlockT> Eq for AlephData<B> {}
+
+impl<B: BlockT> AlephData<B> {
+    pub(crate) fn is_correct(&self) -> bool {
+        match self {
+            AlephData::Empty => true,
+            AlephData::HeadProposal(proposal) => proposal.is_correct(),
+        }
+    }
+}
+
+#[derive(Clone, Debug, Encode, Decode)]
+pub struct AlephProposal<B: BlockT> {
+    // We expect that honest nodes create AlephData {branch: [h_0, h_1, ..., h_n], number: num} objects that represent
+    // an ascending sequence of blocks b_0, b_1, ..., b_n satisfying the following conditions:
+    //     1) hash(b_i) = h_i for i = 0, 1, ..., n,
+    //     2) parent(b_{i+1}) = b_i for i = 0, 1, ..., (n-1),
+    //     3) height(b_n) = num,
+    //     4) The parent of b_0 has been finalized (prior to creating this AlephData).
+    // Such an AlephData object should be thought of as a proposal for block b_n to be finalized.
+    pub branch: Vec<B::Hash>,
+    pub number: NumberFor<B>,
+}
+
+// Need to be implemented manually, as deriving does not work (`BlockT` is not `Hash`).
+impl<B: BlockT> Hash for AlephProposal<B> {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.branch.hash(state);
+        self.number.hash(state);
+    }
+}
+
+// Clippy does not allow deriving PartialEq when implementing Hash manually
+impl<B: BlockT> PartialEq for AlephProposal<B> {
+    fn eq(&self, other: &Self) -> bool {
+        self.number.eq(&other.number) && self.branch.eq(&other.branch)
+    }
+}
+
+impl<B: BlockT> Eq for AlephProposal<B> {}
+
+impl<B: BlockT> AlephProposal<B> {
+    pub(crate) fn new(branch: Vec<B::Hash>, block_number: NumberFor<B>) -> Self {
+        AlephProposal {
+            branch,
+            number: block_number,
+        }
+    }
+
+    pub(crate) fn is_correct(&self) -> bool {
```

Comment:
it's a nit, but this method doesn't guarantee that the proposal satisfies all the assumptions (e.g. parent-child relations), so the 'correctness' in method name can be a little bit misleading

Reply:
I agree, changed it to `is_within_bounds` and added a comment.